{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "0"
      ]
     },
     "execution_count": 1,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "import sys \n",
    "import numpy as np\n",
    "import math\n",
    "import numba\n",
    "from numba import cuda, int64, jit\n",
    "from numba.cuda import random as rnd\n",
    "from timeit import default_timer as timer\n",
    "from scipy.stats import binom, poisson\n",
    "import pickle as pkl\n",
    "import os\n",
    "os.system(\"ml CUDA\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Here we have a little helper function for transforming our binary encoded genomes into a more human readable version."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "@jit\n",
    "def transformAlignment(genotypes):\n",
    "    N=int(genotypes.shape[0])\n",
    "    L=int(genotypes.shape[1])\n",
    "    alignment=[]\n",
    "    for pos in range(0,L*64,2):\n",
    "        posInt = pos//64\n",
    "        posInInt = pos%64\n",
    "        variant=[]\n",
    "        for i in range(N):\n",
    "            if int(genotypes[i,posInt]) & 2**posInInt == 0 and int(genotypes[i,posInt]) & 2**(posInInt+1) == 0:\n",
    "                 variant.append(\"A\")\n",
    "            elif int(genotypes[i,posInt]) & 2**posInInt == 0 and int(genotypes[i,posInt]) & 2**(posInInt+1) > 0:\n",
    "                 variant.append(\"C\")\n",
    "            elif int(genotypes[i,posInt]) & 2**posInInt > 0 and int(genotypes[i,posInt]) & 2**(posInInt+1) == 0:\n",
    "                 variant.append(\"G\")\n",
    "            else:\n",
    "                 variant.append(\"T\")\n",
    "        alignment.append(variant)\n",
    "    return(alignment)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "@cuda.jit\n",
    "def sumF(fitnessarr,F):\n",
    "    F[0]=0\n",
    "    for i in range(fitnessarr.shape[0]):\n",
    "        F[0]+=fitnessarr[i]"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Let's see how we can use ints and bit operations to store and manipulate large amounts of data on a GPU. Our genome is made up of letters of a 4 letter alphabet (ACTG). Most efficient storage is using two bits per letter. For mutations we would like to get a transition:transversion ratio of 1:3. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "@cuda.jit\n",
    "def mutate(rng_states,gen,genotypes,reference,fitnessarr,F,s):\n",
    "    cellID = cuda.grid(1)\n",
    "    genoblocks = genotypes.shape[1]/2\n",
    "    L=64*genoblocks\n",
    "    N=genotypes.shape[0]\n",
    "    targetoffset = 0\n",
    "    if cellID < N:\n",
    "        #add mutations\n",
    "        numberMutations=5\n",
    "        for i in range(numberMutations):\n",
    "            #for simplicity and demonstration purposes we cluster all mutations in generation gen in the genth int (bitblock) and mutate the first five letters\n",
    "            posInt = cellID\n",
    "            posInInt = 2*(5*gen+i)%32\n",
    "            #this will give a transition:transversion ratio of 1:3 on average\n",
    "            r=((cellID+i*N/4)%N)/N\n",
    "            if r < 1/4:\n",
    "                #transverion (first bit with probability 1/2)\n",
    "                if r < 1/8:\n",
    "                    genotypes[cellID,posInt+targetoffset]=genotypes[cellID,posInt+targetoffset] ^ 2**posInInt\n",
    "                #second bit has to be flipped\n",
    "                genotypes[cellID,posInt+targetoffset]=genotypes[cellID,posInt+targetoffset] ^ 2**(posInInt+1)\n",
    "            else:\n",
    "                #transition (00<->10 or 01<->11)\n",
    "                genotypes[cellID,posInt+targetoffset]=genotypes[cellID,posInt+targetoffset] ^ 2**posInInt\n",
    "        #not we are gonna count the mutatations, i.e., mismatches with the reference\n",
    "        snpCount=0\n",
    "        for posInt in range(genoblocks):\n",
    "            intAnd=(genotypes[cellID,posInt+targetoffset] ^ reference[posInt])\n",
    "            for posInInt in range(0,64,2):\n",
    "                if intAnd & 2**posInInt > 0 or intAnd & 2**(posInInt+1) > 0:\n",
    "                    snpCount+=1\n",
    "        fitnessarr[cellID]=s**snpCount\n",
    "    cuda.syncthreads()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [],
   "source": [
    "L=10000\n",
    "n=64*L\n",
    "\n",
    "threadsperblock = 20\n",
    "blocks = 20\n",
    "\n",
    "N=threadsperblock*blocks\n",
    "s=0.999\n",
    "SEED=1\n",
    "NUMCYCLES=8*N\n",
    "\n",
    "genotypes = np.array(np.reshape(np.array(list(np.random.randint(0,2**64, L*2,dtype=np.uint64))*2*N),(N,L*4)), dtype=np.uint64)\n",
    "devgenotypes = cuda.to_device(genotypes)\n",
    "reference = np.copy(genotypes[0,:])\n",
    "devreference = cuda.to_device(reference)\n",
    "\n",
    "fitnessarr = np.ones(N)\n",
    "F=np.sum(fitnessarr)\n",
    "devfitnessarr=cuda.to_device(fitnessarr)\n",
    "rng_states = rnd.create_xoroshiro128p_states(threadsperblock * blocks, seed=SEED)\n",
    "F=cuda.device_array((1))\n",
    "sumF[1,1](devfitnessarr,F)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "first iteration: 0.35143856797367334\n",
      "[0.99500999 0.99500999 0.99500999 0.99500999 0.99500999 0.99500999\n",
      " 0.99500999 0.99500999 0.99500999 0.99500999]\n",
      "GATTTACAATTAAAGTACGCTAGTTCTAAGCACACCAGGCATCAGATTTCATAAGTGTGTGGACTGATTCTGGGCGGACTGTCTTTCAGGACTAGA\n",
      "CGCCAACAATTAAAGTACGCTAGTTCTAAGCAGGTTTGGCATCAGATTTCATAAGTGTGTGGACTGATTCTGGGCGGACTGTCTTTCAGGACTAGA\n",
      "CGCCAACAATTAAAGTACGCTAGTTCTAAGCACACCAGGCATCAGATTTCATAAGTGTGTGGACAAGCACTGGGCGGACTGTCTTTCAGGACTAGA\n",
      "CGCCAACAATTAAAGTACGCTAGTTCTAAGCACACCAGGCATCAGATTTCATAAGTGTGTGGACTGATTCTGGGCGGACTGTCTTTCAGGACTAGA\n",
      "CGCCAACAATTAAAGTACGCTAGTTCTAAGCACACCAGGCATCAGATTTCATAAGTGTGTGGACTGATTCTGGGCGGACTGTCTTTCAGGACTAGA\n",
      "CGCCAACAATTAAAGTACGCTAGTTCTAAGCACACCAGGCATCAGATTTCATAAGTGTGTGGACTGATTCTGGGCGGACTGTCTTTCAGGACTAGA\n",
      "CGCCAACAATTAAAGTACGCTAGTTCTAAGCACACCAGGCATCAGATTTCATAAGTGTGTGGACTGATTCTGGGCGGACTGTCTTTCAGGACTAGA\n",
      "CGCCAACAATTAAAGTACGCTAGTTCTAAGCACACCAGGCATCAGATTTCATAAGTGTGTGGACTGATTCTGGGCGGACTGTCTTTCAGGACTAGA\n",
      "CGCCAACAATTAAAGTACGCTAGTTCTAAGCACACCAGGCATCAGATTTCATAAGTGTGTGGACTGATTCTGGGCGGACTGTCTTTCAGGACTAGA\n",
      "CGCCAACAATTAAAGTACGCTAGTTCTAAGCACACCAGGCATCAGATTTCATAAGTGTGTGGACTGATTCTGGGCGGACTGTCTTTCAGGACTAGA\n",
      "CGCCAACAATTAAAGTACGCTAGTTCTAAGCACACCAGGCATCAGATTTCATAAGTGTGTGGACTGATTCTGGGCGGACTGTCTTTCAGGACTAGA\n",
      "CGCCAACAATTAAAGTACGCTAGTTCTAAGCACACCAGGCATCAGATTTCATAAGTGTGTGGACTGATTCTGGGCGGACTGTCTTTCAGGACTAGA\n",
      "CGCCAACAATTAAAGTACGCTAGTTCTAAGCACACCAGGCATCAGATTTCATAAGTGTGTGGACTGATTCTGGGCGGACTGTCTTTCAGGACTAGA\n",
      "CGCCAACAATTAAAGTACGCTAGTTCTAAGCACACCAGGCATCAGATTTCATAAGTGTGTGGACTGATTCTGGGCGGACTGTCTTTCAGGACTAGA\n",
      "CGCCAACAATTAAAGTACGCTAGTTCTAAGCACACCAGGCATCAGATTTCATAAGTGTGTGGACTGATTCTGGGCGGACTGTCTTTCAGGACTAGA\n",
      "CGCCAACAATTAAAGTACGCTAGTTCTAAGCACACCAGGCATCAGATTTCATAAGTGTGTGGACTGATTCTGGGCGGACTGTCTTTCAGGACTAGA\n",
      "CGCCAACAATTAAAGTACGCTAGTTCTAAGCACACCAGGCATCAGATTTCATAAGTGTGTGGACTGATTCTGGGCGGACTGTCTTTCAGGACTAGA\n",
      "CGCCAACAATTAAAGTACGCTAGTTCTAAGCACACCAGGCATCAGATTTCATAAGTGTGTGGACTGATTCTGGGCGGACTGTCTTTCAGGACTAGA\n",
      "CGCCAACAATTAAAGTACGCTAGTTCTAAGCACACCAGGCATCAGATTTCATAAGTGTGTGGACTGATTCTGGGCGGACTGTCTTTCAGGACTAGA\n",
      "CGCCAACAATTAAAGTACGCTAGTTCTAAGCACACCAGGCATCAGATTTCATAAGTGTGTGGACTGATTCTGGGCGGACTGTCTTTCAGGACTAGA\n"
     ]
    }
   ],
   "source": [
    "start = timer()\n",
    "mutate[blocks, threadsperblock](rng_states,0,devgenotypes,devreference,devfitnessarr,F,s)\n",
    "sumF[1,1](devfitnessarr,F)\n",
    "evolve_time=timer()-start\n",
    "print(\"first iteration: \"+str(evolve_time))\n",
    "fitnessarr = devfitnessarr.copy_to_host()\n",
    "print(fitnessarr[:10])\n",
    "genotypes = devgenotypes.copy_to_host()\n",
    "alignment = transformAlignment(genotypes[:20,:3])\n",
    "x=[print(\"\".join(row)) for row in np.array(alignment).T]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "second iteration: 0.0010627079755067825\n",
      "[0.99004488 0.99004488 0.99004488 0.99004488 0.99004488 0.99004488\n",
      " 0.99004488 0.99004488 0.99004488 0.99004488]\n",
      "GATTTTTGGATAAAGTACGCTAGTTCTAAGCACACCAGGCATCAGATTTCATAAGTGTGTGGACTGATTCTGGGCGGACTGTCTTTCAGGACTAGA\n",
      "CGCCAACAATTAAAGTACGCTAGTTCTAAGCAGGTTTCATGACAGATTTCATAAGTGTGTGGACTGATTCTGGGCGGACTGTCTTTCAGGACTAGA\n",
      "CGCCAACAATTAAAGTACGCTAGTTCTAAGCACACCAGGCATCAGATTTCATAAGTGTGTGGACAAGCAGCAACCGGACTGTCTTTCAGGACTAGA\n",
      "CGCCAACAATTAAAGTACGCTAGTTCTAAGCACACCAGGCATCAGATTTCATAAGTGTGTGGACTGATTCTGGGCGGACTGTCTTTCAGGACTAGA\n",
      "CGCCAACAATTAAAGTACGCTAGTTCTAAGCACACCAGGCATCAGATTTCATAAGTGTGTGGACTGATTCTGGGCGGACTGTCTTTCAGGACTAGA\n",
      "CGCCAACAATTAAAGTACGCTAGTTCTAAGCACACCAGGCATCAGATTTCATAAGTGTGTGGACTGATTCTGGGCGGACTGTCTTTCAGGACTAGA\n",
      "CGCCAACAATTAAAGTACGCTAGTTCTAAGCACACCAGGCATCAGATTTCATAAGTGTGTGGACTGATTCTGGGCGGACTGTCTTTCAGGACTAGA\n",
      "CGCCAACAATTAAAGTACGCTAGTTCTAAGCACACCAGGCATCAGATTTCATAAGTGTGTGGACTGATTCTGGGCGGACTGTCTTTCAGGACTAGA\n",
      "CGCCAACAATTAAAGTACGCTAGTTCTAAGCACACCAGGCATCAGATTTCATAAGTGTGTGGACTGATTCTGGGCGGACTGTCTTTCAGGACTAGA\n",
      "CGCCAACAATTAAAGTACGCTAGTTCTAAGCACACCAGGCATCAGATTTCATAAGTGTGTGGACTGATTCTGGGCGGACTGTCTTTCAGGACTAGA\n",
      "CGCCAACAATTAAAGTACGCTAGTTCTAAGCACACCAGGCATCAGATTTCATAAGTGTGTGGACTGATTCTGGGCGGACTGTCTTTCAGGACTAGA\n",
      "CGCCAACAATTAAAGTACGCTAGTTCTAAGCACACCAGGCATCAGATTTCATAAGTGTGTGGACTGATTCTGGGCGGACTGTCTTTCAGGACTAGA\n",
      "CGCCAACAATTAAAGTACGCTAGTTCTAAGCACACCAGGCATCAGATTTCATAAGTGTGTGGACTGATTCTGGGCGGACTGTCTTTCAGGACTAGA\n",
      "CGCCAACAATTAAAGTACGCTAGTTCTAAGCACACCAGGCATCAGATTTCATAAGTGTGTGGACTGATTCTGGGCGGACTGTCTTTCAGGACTAGA\n",
      "CGCCAACAATTAAAGTACGCTAGTTCTAAGCACACCAGGCATCAGATTTCATAAGTGTGTGGACTGATTCTGGGCGGACTGTCTTTCAGGACTAGA\n",
      "CGCCAACAATTAAAGTACGCTAGTTCTAAGCACACCAGGCATCAGATTTCATAAGTGTGTGGACTGATTCTGGGCGGACTGTCTTTCAGGACTAGA\n",
      "CGCCAACAATTAAAGTACGCTAGTTCTAAGCACACCAGGCATCAGATTTCATAAGTGTGTGGACTGATTCTGGGCGGACTGTCTTTCAGGACTAGA\n",
      "CGCCAACAATTAAAGTACGCTAGTTCTAAGCACACCAGGCATCAGATTTCATAAGTGTGTGGACTGATTCTGGGCGGACTGTCTTTCAGGACTAGA\n",
      "CGCCAACAATTAAAGTACGCTAGTTCTAAGCACACCAGGCATCAGATTTCATAAGTGTGTGGACTGATTCTGGGCGGACTGTCTTTCAGGACTAGA\n",
      "CGCCAACAATTAAAGTACGCTAGTTCTAAGCACACCAGGCATCAGATTTCATAAGTGTGTGGACTGATTCTGGGCGGACTGTCTTTCAGGACTAGA\n"
     ]
    }
   ],
   "source": [
    "start = timer()\n",
    "mutate[blocks, threadsperblock](rng_states,1,devgenotypes,devreference,devfitnessarr,F,s)\n",
    "sumF[1,1](devfitnessarr,F)\n",
    "evolve_time=timer()-start\n",
    "print(\"second iteration: \"+str(evolve_time))\n",
    "fitnessarr = devfitnessarr.copy_to_host()\n",
    "print(fitnessarr[:10])\n",
    "genotypes = devgenotypes.copy_to_host()\n",
    "alignment = transformAlignment(genotypes[:20,:3])\n",
    "x=[print(\"\".join(row)) for row in np.array(alignment).T]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "100 more iterations: 0.04910828825086355\n",
      "[0.99202794 0.99202794 0.99202794 0.99202794 0.99202794 0.99202794\n",
      " 0.99202794 0.99202794 0.99202794 0.99202794]\n",
      "AGCAACCACTGAACCAACGCTAGTTCTAAGCACACCAGGCATCAGATTTCATAAGTGTGTGGACTGATTCTGGGCGGACTGTCTTTCAGGACTAGA\n",
      "CGCCAACAATTAAAGTACGCTAGTTCTAAGCAAACAATGCCTAAGCAATCATAAGTGTGTGGACTGATTCTGGGCGGACTGTCTTTCAGGACTAGA\n",
      "CGCCAACAATTAAAGTACGCTAGTTCTAAGCACACCAGGCATCAGATTTCATAAGTGTGTGGACGGAGTATGTGAGGCGAGTCTTTCAGGACTAGA\n",
      "CGCCAACAATTAAAGTACGCTAGTTCTAAGCACACCAGGCATCAGATTTCATAAGTGTGTGGACTGATTCTGGGCGGACTGTCTTTCAGGACTAGA\n",
      "CGCCAACAATTAAAGTACGCTAGTTCTAAGCACACCAGGCATCAGATTTCATAAGTGTGTGGACTGATTCTGGGCGGACTGTCTTTCAGGACTAGA\n",
      "CGCCAACAATTAAAGTACGCTAGTTCTAAGCACACCAGGCATCAGATTTCATAAGTGTGTGGACTGATTCTGGGCGGACTGTCTTTCAGGACTAGA\n",
      "CGCCAACAATTAAAGTACGCTAGTTCTAAGCACACCAGGCATCAGATTTCATAAGTGTGTGGACTGATTCTGGGCGGACTGTCTTTCAGGACTAGA\n",
      "CGCCAACAATTAAAGTACGCTAGTTCTAAGCACACCAGGCATCAGATTTCATAAGTGTGTGGACTGATTCTGGGCGGACTGTCTTTCAGGACTAGA\n",
      "CGCCAACAATTAAAGTACGCTAGTTCTAAGCACACCAGGCATCAGATTTCATAAGTGTGTGGACTGATTCTGGGCGGACTGTCTTTCAGGACTAGA\n",
      "CGCCAACAATTAAAGTACGCTAGTTCTAAGCACACCAGGCATCAGATTTCATAAGTGTGTGGACTGATTCTGGGCGGACTGTCTTTCAGGACTAGA\n",
      "CGCCAACAATTAAAGTACGCTAGTTCTAAGCACACCAGGCATCAGATTTCATAAGTGTGTGGACTGATTCTGGGCGGACTGTCTTTCAGGACTAGA\n",
      "CGCCAACAATTAAAGTACGCTAGTTCTAAGCACACCAGGCATCAGATTTCATAAGTGTGTGGACTGATTCTGGGCGGACTGTCTTTCAGGACTAGA\n",
      "CGCCAACAATTAAAGTACGCTAGTTCTAAGCACACCAGGCATCAGATTTCATAAGTGTGTGGACTGATTCTGGGCGGACTGTCTTTCAGGACTAGA\n",
      "CGCCAACAATTAAAGTACGCTAGTTCTAAGCACACCAGGCATCAGATTTCATAAGTGTGTGGACTGATTCTGGGCGGACTGTCTTTCAGGACTAGA\n",
      "CGCCAACAATTAAAGTACGCTAGTTCTAAGCACACCAGGCATCAGATTTCATAAGTGTGTGGACTGATTCTGGGCGGACTGTCTTTCAGGACTAGA\n",
      "CGCCAACAATTAAAGTACGCTAGTTCTAAGCACACCAGGCATCAGATTTCATAAGTGTGTGGACTGATTCTGGGCGGACTGTCTTTCAGGACTAGA\n",
      "CGCCAACAATTAAAGTACGCTAGTTCTAAGCACACCAGGCATCAGATTTCATAAGTGTGTGGACTGATTCTGGGCGGACTGTCTTTCAGGACTAGA\n",
      "CGCCAACAATTAAAGTACGCTAGTTCTAAGCACACCAGGCATCAGATTTCATAAGTGTGTGGACTGATTCTGGGCGGACTGTCTTTCAGGACTAGA\n",
      "CGCCAACAATTAAAGTACGCTAGTTCTAAGCACACCAGGCATCAGATTTCATAAGTGTGTGGACTGATTCTGGGCGGACTGTCTTTCAGGACTAGA\n",
      "CGCCAACAATTAAAGTACGCTAGTTCTAAGCACACCAGGCATCAGATTTCATAAGTGTGTGGACTGATTCTGGGCGGACTGTCTTTCAGGACTAGA\n"
     ]
    }
   ],
   "source": [
    "start = timer()\n",
    "for i in range(2,102):\n",
    "    mutate[blocks, threadsperblock](rng_states,i,devgenotypes,devreference,devfitnessarr,F,s)\n",
    "    sumF[1,1](devfitnessarr,F)\n",
    "evolve_time=timer()-start\n",
    "print(\"100 more iterations: \"+str(evolve_time))\n",
    "fitnessarr = devfitnessarr.copy_to_host()\n",
    "print(fitnessarr[:10])\n",
    "genotypes = devgenotypes.copy_to_host()\n",
    "alignment = transformAlignment(genotypes[:20,:3])\n",
    "x=[print(\"\".join(row)) for row in np.array(alignment).T]"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.6.8"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
